collision_detection_fcl/src/collision_common.cpp

```diff
@@ -455,6 +455,133 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
+  const CollisionGeometryData* cd2 = static_cast<const CollisionGeometryData*>(o2->collisionGeometry()->getUserData());
+
+  // If active components are specified
+  if (cdata->active_components_only_)
+  {
+    const robot_model::LinkModel *l1 = cd1->type == BodyTypes::ROBOT_LINK ? cd1->ptr.link : (cd1->type == BodyTypes::ROBOT_ATTACHED ? cd1->ptr.ab->getAttachedLink() : NULL);
+    const robot_model::LinkModel *l2 = cd2->type == BodyTypes::ROBOT_LINK ? cd2->ptr.link : (cd2->type == BodyTypes::ROBOT_ATTACHED ? cd2->ptr.ab->getAttachedLink() : NULL);
+
+    // If neither of the involved components is active
+    if ((!l1 || cdata->active_components_only_->find(l1) == cdata->active_components_only_->end()) &&
+        (!l2 || cdata->active_components_only_->find(l2) == cdata->active_components_only_->end()))
+    {
+      return false;
+    }
+  }
+
+  // use the collision matrix (if any) to avoid certain distance checks
+  bool always_allow_collision = false;
+  if (cdata->acm_)
+  {
+    AllowedCollision::Type type;
+
+    bool found = cdata->acm_->getAllowedCollision(cd1->getID(), cd2->getID(), type);
+    if (found)
+    {
+      // if we have an entry in the collision matrix, we read it
+      if (type == AllowedCollision::ALWAYS)
+      {
+        always_allow_collision = true;
+        if (!cdata->req_->verbose)
+          logDebug("Collision between '%s' and '%s' is always allowed. No contacts are computed.",
+                   cd1->getID().c_str(), cd2->getID().c_str());
+      }
+    }
+  }
+
+  // check if a link is touching an attached object
+  if (cd1->type == BodyTypes::ROBOT_LINK && cd2->type == BodyTypes::ROBOT_ATTACHED)
+  {
+    const std::set<std::string> &tl = cd2->ptr.ab->getTouchLinks();
+    if (tl.find(cd1->getID()) != tl.end())
+    {
+      always_allow_collision = true;
+      if (!cdata->req_->verbose)
+        logDebug("Robot link '%s' is allowed to touch attached object '%s'. No contacts are computed.",
+                 cd1->getID().c_str(), cd2->getID().c_str());
+    }
+  }
+  else
+  {
+    if (cd2->type == BodyTypes::ROBOT_LINK && cd1->type == BodyTypes::ROBOT_ATTACHED)
+    {
+      const std::set<std::string> &tl = cd1->ptr.ab->getTouchLinks();
+      if (tl.find(cd2->getID()) != tl.end())
+      {
+        always_allow_collision = true;
+        if (!cdata->req_->verbose)
+          logDebug("Robot link '%s' is allowed to touch attached object '%s'. No contacts are computed.",
+                   cd2->getID().c_str(), cd1->getID().c_str());
+      }
+    }
+  }
+
+  if(always_allow_collision)
+  {
+    return false;
+  }
+
+  if (!cdata->req_->verbose)
+    logDebug("Actually checking collisions between %s and %s", cd1->getID().c_str(), cd2->getID().c_str());
+
+  fcl::DistanceResult dist_result;
+  double d = fcl::distance(o1, o2, fcl::DistanceRequest(true), dist_result);
+
+  // Store detailed distance information for each objects
```

Comment:
This is a bit odd.  The method above `fcl::distance` returns the same thing as `distanceDetailedCallback`.  Everything below here appears to cache the result?  How is the `distance_detailed` used beyond this point?  If `distanceDetailedCallback` is not called for every object pair, then the results in `distance_detailed` are incomplete (am I correct?).  Does the calling object/program inherently know this?